core/http/server-spring/src/main/java/org/eclipse/rdf4j/http/server/repository/transaction/Transaction.java

```diff
@@ -134,13 +137,12 @@ void begin(IsolationLevel level)
 	void rollback()
 		throws InterruptedException, ExecutionException
 	{
-		Future<Boolean> result;
-		synchronized (futures) {
-			result = executor.submit(() -> {
-				txnConnection.rollback();
-				return true;
-			});
+		Future<Boolean> result = executor.submit(() -> {
+			txnConnection.rollback();
+			return true;
+		});
 
+		synchronized (futures) {
```

Comment:
I have added a Semaphore to make task executions mutually exclusive of the hasActiveOperations method

I have readded the synchronisation for the futures back around the ExecutorService.submit method to make ExecutorService.submit and futures.add atomic from the perspective of hasActiveOperations.

I am a little concerned that hasActiveOperations and the other transaction methods do not take into account the closed status to abort early if the Transaction has already closed up shop. Ie, they currently need to obtain the monitor (increasing the likelihood of deadlocks) before being given an ExecutionException to indicate that nothing more can occur, once ExecutorService.shutdown has been called in the close method. I have added two AtomicBooleans to track the start and end of the closed method, but they are not linked to hasActiveOperations or ActiveTransactionRegistry yet.